### ISSUE TITLE
Scout/Scout Extended integration

### ISSUE THREAD

**Issue by lukecurtis93** [spatie/laravel-query-builder#147]
Is there currently a way to integrate using Laravel scout or [Scout extended](https://github.com/algolia/scout-extended) in the Query builder?

As these use custom builders I've been unable to get a raw builder for these objects in the query.

If anyone could point me in the right direction that'd be super helpful

**Comment by freekmurze** [comment]
All I can say is that you can just pass a `Builder` instance to the constructor of `QueryBuilder`. Unfortunately I don't know how to get a `Builder` instance out of Scout. 

Check the docs of Scout or ask the question on the forum at Laracasts.

**Comment by msassa** [comment]
@lukecurtis93 Did you find any way to use this package with Scout?

**Comment by lukecurtis93** [comment]
Unfortunately not - I've opened a thread here, https://laracasts.com/discuss/channels/laravel/get-builder-instance-from-laravel-scout and and issue here https://github.com/laravel/scout/issues/343

But nothing yet.

**Comment by ndrkltsk** [comment]
Had the same issue, solved it in this way:

Get the ids of the models that satisfies the research:
``` $models = Model::search(request()->q)->get()->pluck('id') ```

and concatenate a whereIn to the QueryBuilder chain:
```QueryBuilder::for(Model::class)->whereIn('id', $models)```

It's a bit ugly but works, and the performance is ok

**Comment by michgeek** [comment]
This one would allow you to paginate and reuse the query

```php
$query = QueryBuilder::for(Model::class)->getQuery();

$result = Model::search(request()->q, function ($q) use ($query) {
    return $query->mergeWheres($q->getQuery()->wheres, $q->getBindings());
})->paginate();
```

**Comment by Jackardios** [comment]
For such cases, I made a separate version from the sources of this project, which can be used separately from the query builder.

**Comment by lukecurtis93** [comment]
Phew this is a blast from the past...

But I think the answer is pretty conclusive here. No it's not possible to do anything but the suggestion above afaik. The builder from scout is [too different](https://github.com/laravel/scout/issues/343#issuecomment-453695591) to support something like this.

In the years that have gone by I'd probably come to the personal conclusion now that query builder/scout serve two different use cases to be completely honest for what I was originally attempting to do.

I'd probably lean to using [facets](https://www.algolia.com/doc/guides/managing-results/refine-results/filtering/#how-to-filter-your-data) if I was to revisit this problem again (it was a public-esque facing thing).

If your use case is more of an internal/employee application-esque tool I'd definitely say you could leverage this query builder package and have something like [this](https://freek.dev/1182-searching-models-using-a-where-like-query-in-laravel) available under the hood for a custom scope. (YMMV for scale)

Just my 2c

**Comment by SlimGee** [comment]
> Is this the best option for the moment?

After looking through the comments I came with relatively cleaner solution based on @ndrkltsk 's comment

in your model add a scope

```php
/**
 * Scope a query to only include models that are returned by scout
 *
 * @param Builder $query
 * @param string $search
 * @return Builder
 */
public function scopeWhereScout(Builder $query, string $search): Builder
{
    return $query->whereIn(
        'id',
        self::search($search)
            ->get()
            ->pluck('id'),
    );
}
```

then use a `scopeFilter` like this

```php
QueryBuilder::for(YourModel::class) ->allowedFilters([AllowedFilter::scope('search', 'whereScout')])
```

**Comment by CodeWrap** [comment]
For those who need to keep the order of the items:
```
$ids = self::search($search)
    ->get()
    ->pluck('id')
    ->toArray();

$orders = array_map(function ($item) {
    return "id = {$item} desc";
}, $ids);
$rawOrder = implode(', ', $orders);

return $query->whereIn('id', $ids)->orderByRaw($rawOrder);
```

**Comment by der-Lehmann** [comment]
As far as I can see, there is one caveat with the mentioned approach.
That is because the search with scout will only return the first 20 models. When the eloquent search does not include those 20 models, no results are shown at all.

Maybe this behaviour is different when using a driver other than meilisearch. But for me I was only getting the first 20 items returned by the `search` method of scout while using meilisearch.

In order to solve this, I modified the approach mentioned above like this:

```php
$builder = self::search($value);
$builder->limit = 10000000;
$result = $builder->raw();
$ids = array_column($result['hits'], 'id');

if (count($ids)) {
    $orders = array_map(fn($id) => sprintf("id = %d desc", $id), $ids);
    $rawOrder = implode(', ', $orders);
    $query->whereIn('id', $ids);
    $query->orderByRaw($rawOrder);
}
```

**Comment by francoism90** [comment]
In case it's useful for anyone (@ianfortier): https://github.com/foxws/laravel-scout-builder

I have created a fork of laravel-query-builder, and convert it so it works with Laravel Scout.

It's still beta quality, and it would be great to actually merge parts into this package instead, or @freekmurze / spatie has interests in creating something like it themselves. :)

All credits to Spatie, the only thing I did, is the conversion with help of AI to speed up this process.
